uat/driver/test_bundle.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+import os
+from pathlib import Path
+
+import pytest
+from lightkube import ApiError, Client, codecs
+from lightkube.generic_resource import create_global_resource, load_in_cluster_generic_resources
+from utils import assert_namespace_active, delete_job, fetch_job_logs, wait_for_job
+
+log = logging.getLogger(__name__)
+
+ASSETS_DIR = Path("assets")
+JOB_TEMPLATE_FILE = ASSETS_DIR / "test-job.yaml.j2"
+PROFILE_TEMPLATE_FILE = ASSETS_DIR / "test-profile.yaml.j2"
+
+TESTS_DIR = os.path.abspath(Path("integrations"))
+TESTS_IMAGE = "kubeflownotebookswg/jupyter-scipy:v1.7.0"
+
+NAMESPACE = "test-kubeflow"
+PROFILE_RESOURCE = create_global_resource(
+    group="kubeflow.org",
+    version="v1",
+    kind="profile",
+    plural="profiles",
+)
+
+JOB_NAME = "test-kubeflow"
+
+
+@pytest.fixture(scope="module")
+def lightkube_client():
+    """Initialise Lightkube Client."""
+    lightkube_client = Client()
+    load_in_cluster_generic_resources(lightkube_client)
+    return lightkube_client
+
+
+@pytest.fixture(scope="module")
+def create_profile(lightkube_client):
+    """Create Profile and handle cleanup at the end of the module tests."""
+    log.info(f"Creating Profile {NAMESPACE}...")
+    resources = list(
+        codecs.load_all_yaml(
+            PROFILE_TEMPLATE_FILE.read_text(),
+            context={"namespace": NAMESPACE},
+        )
+    )
+    assert len(resources) == 1, f"Expected 1 Profile, got {len(resources)}!"
+    lightkube_client.create(resources[0])
+
+    yield
+
+    # delete the Profile at the end of the module tests
+    log.info(f"Deleting Profile {NAMESPACE}...")
+    lightkube_client.delete(PROFILE_RESOURCE, name=NAMESPACE)
+
+
+@pytest.mark.abort_on_fail
+async def test_create_profile(lightkube_client, create_profile):
+    """Test Profile creation.
+
+    This test relies on the create_profile fixture, which handles the Profile creation and
+    is responsible for cleaning up at the end.
+    """
+    try:
+        profile_created = lightkube_client.get(
+            PROFILE_RESOURCE,
+            name=NAMESPACE,
+        )
+    except ApiError as e:
+        if e.status == 404:
+            profile_created = False
+        else:
+            raise
+    assert profile_created, f"Profile {NAMESPACE} not found!"
+
+    assert_namespace_active(lightkube_client, NAMESPACE)
+
+
+def test_kubeflow(lightkube_client):
+    """Run a K8s Job to execute the notebook tests."""
```

Comment:
Not a showstopper, but ideally (as discussed previously) it would be nice if each workload test showed up here (maybe all as separate jobs) rather than was nested.  It is still readable as is, just a little unexpected/opaque

Reply:
I feel that if we did that, we would shift the responsibility for test discovery from the standalone `integrations` to the generic driver. Then, I'm not sure `integrations` would be standalone anymore.

The point is to be able to run that directory directly from inside a Notebook, and therefore it needs to contain the logic about "collecting" the tests and automating their execution. I'm not sure that the driver can know anything about these with the above in mind.

An alternative would be to "duplicate" the test discovery and execution logic in the `driver` so that when running the tests on K8s Jobs we don't go through `test_integrations.py`. Then I guess we could have a separate Job for each test. I'm not sure I like this TBH, I'd rather invest in enhancing the existing solution to stream the Job logs so that we know what's happening at any point in time.

Reply:
yeah I think keeping it as is for now sounds good.  Worst case, we address it in future if we find the current state hard to interpret in CI

Reply:
Thinking a little more on this, having timeouts per integration might be the real selling point here.  If the first integration hangs, nothing else runs and it looks like they're all broken.  But again, good to think about another day